fm/qtfm.1
.TH qtfm 1 "01 August 2018" "6.2" "QtFM documentation"
.SH NAME
QtFM -\ desktop independent file manager
.SH SYNOPSIS
.B qtfm
.I file
.B ...
.SH DESCRIPTION
Lightweight file manager using Qt.
.SH FILES
.I ~/.config/qtfm6/qtfm6.conf
.RS
Per user configuration file.
.RE
.I ~/.config/qtfm6/file.cache
.RS
Per user file/mime icon cache.
.RE
.I ~/.config/qtfm6/folder.cache
.RS
Per user folder icon cache. Currently not used.
.RE
.I ~/.config/qtfm6/thumbs.cache
.RS
Per user thumbnail icon cache.
.RE
.SH SEE ALSO
qtfm-tray(1)
.SH BUGS
See https://github.com/rodlie/qtfm/issues
.SH AUTHOR
QtFM developers


ChangeLog
6.1.7 11-February 2019
- context menu fixes
- settings dialog: fix performance issues on old CPUs

6.1.6 31-January 2019
- progress dialog fix
- path tab-completion. Thanks to @kubaroth for the patch
- completer handles expansion of ~ path. Thanks to @kubaroth for the patch
- support mouse back button
- macOS support
- NetBSD support

6.1.5 09-December 2018
- fix build if CONFIG+=no_dbus enabled
- fixed broken multiple file copy/move instances

6.1.4 26-November 2018
- fix potential config corruption. Thanks to @slackuser0xae34 for the patch
- fix segfault on thumbnail update (view may have changed since loading thumbnails)
- don't cut file(s) if moving to same directory in bookmarks
- verify reply from udisks (avoid crash)
- populate view if not watched
- fix refresh view
- improved performance

6.1.3 07-October 2018
- update grid on paste/remove
- settings: don't show "launcher" in application dialog
- added manual refresh of path

6.1.2 11-September 2018
- fixed broken file type check on custom actions
- added new (default) custom actions
- fixed issue with /etc/mtab not being read properly on some systems
- settings: added restore custom actions to default option
- FreeBSD: fixed issue with UDisks
- associate more mime types with icons

6.1.1 25-August 2018
- fixed ask dialog in bookmarks when drag-and-drop
- update listview grid earlier
- tray: fix open mountpoint in qtfm (with spaces)

6.1.0 04-August 2018
- added system tray daemon (qtfm-tray), auto mount/play storage/optical
- drag-and-drop: add settings for modifiers
- inotify fixes
- fixed refresh of thumbnails
- added singleclick options in settings
- several ui fixes/enhancements
- added dbus session support
- various fixes/changes in settings
- various bug fixes
- added new build options
- added (icon) zoom support where missing
- proper support for multiple files in external applications
- advanced settings
- FreeBSD support
- bookmarks drop fixes
- mime fixes

6.0.3 11-June 2018
- fixed mount issue with optical in udisks

6.0.2 08-June 2018
- fixed another bookmark related bug

6.0.1 07-June 2018
- fixed broken bookmarks save
- fixed segfault if udisks not running

6.0.0 04-June 2018
- support for removable devices (udisks)
- basic trash support
- several bug/ui/build/mime fixes
- several UI enhancements/speed up
- improved desktop integration

5.9 25-May 2013
- fixed mime type handling
- fixed problem with opening qpdfview
- fixed application speed up, saving of properties much faster
- fixed missing button in about dialog

5.8 5-May 2013
- added support for Qt5 (application can be compiled with both Qt4 or Qt5)
- added *.desktop file based application selector
- added support for changing icon theme from settings
- added process dialog for custom actions. If custom action is executed, then dialog with action output popups. When action is finished, dialog is closed. This feature can be turned on/off in settings
- added mime type based default application configuration into the settings dialog
- added default mime app configuration file selection to settings (mimeapps.list/defaults.list)
- added default mime apps autogeneration from *.desktop files
- added about box
- modified icon view more mime types are recognized now
- modified more compact settings dialog

5.7 21-December 2012
- added settings manager, almost all settings are now centralized in one dialog
- modified edit menu
- modified custom action management centralized to CustomActionManager
- fixed bug loading of file icons, versioned files like 'xy.so.1.0.0' are now displayed propertly
- fixed bug in bookmarks: newly created bookmarks were not appeared in list of shortcuts until application was restarted or custom actions dialog was opened
- fixed bug in shortcuts: sometimes shortcut editor show duplicated bookmarks

5.6 17-October 2012
- added GUI for drag & drop: qtfm asks user about what he want to do (copy, move or create link)
- added sorting (name, size and date in ascending or descending order)
- added FileUtils in order to provide file management routines that can be reused 
- added main menu section "Go"  
- modified drag & drop; when holding ctrl/shift files are copied/moved 
- modified strings "/" was also replaced with QDir::separator() 

5.5	15-June-2012
- modified file xdg associations to use mimeapps.list.
- modified tab right-click to activate selected tab.
- modified thumbs, uniform size/aspect
- modified 'Refresh' to update mimetype if changed.
- modified xdg configuration to ensure '~/.local/share/applications/defaults.list' exists.
- fixed bug not allowing a cut/move on same device with limited space.
- fixed bug not using custom action override on double-click.
- fixed bug with moving across filesystems.
- fixed bug with multiple custom actions of same type not appearing.

5.4	7-March-2012
- added 'tabsOnTop' option.
- added new detection for automounts.
- added ability to set shortcuts for bookmarks.
- added option to use mimetype instead of suffix in detail mode and custom actions.
- modified filetype for file specific custom actions, now treated like a wildcard.
- modified thumbs to show in all views.
- modified text colours to follow system theme.
- modified status/hover info updating.
- fixed bug in filesystem model not updating in some cases.
- fixed bug not showing hidden items in '/' root.
- fixed bug trying to open a tab on a file.
- fixed bug opening folder from command line (Qt4.8).

5.3	14-December-2011
- added new address bar navigation and drop-down auto-completion.
- added 'New tab' to tree context menu.
- added 'Close' context menu to tabs.
- added Chinese translation (zh).
- fixed 'single-click' navigation mode.
- fixed sorting for size & date columns in detail view.
- fixed bug in available space detection when pasting to removable drives.

5.2	28-November-2011
- complete rewrite of FS model, no longer using QFileSystemModel.
- fixed many small issues that were associated with QFileSystemModel.

fm7/qtfm.cpp
#include "qtfm.h"
#include "common.h"

#include <QMdiSubWindow>

QtFM::QtFM(QWidget *parent)
    : QMainWindow(parent)
    , mdi(Q_NULLPTR)
    , mimes(Q_NULLPTR)
    , mime(true)
    , mBar(Q_NULLPTR)
    , sBar(Q_NULLPTR)
    , navBar(Q_NULLPTR)
    , pathEdit(Q_NULLPTR)
    , fileMenu(Q_NULLPTR)
    , editMenu(Q_NULLPTR)
    , viewMenu(Q_NULLPTR)
    , tileAction(Q_NULLPTR)
    , tabViewAction(Q_NULLPTR)
    , backButton(Q_NULLPTR)
    , upButton(Q_NULLPTR)
    , homeButton(Q_NULLPTR)
    , tileButton(Q_NULLPTR)
{
    QIcon::setThemeSearchPaths(Common::iconPaths(qApp->applicationDirPath()));
    Common::setupIconTheme(qApp->applicationFilePath());

    setWindowIcon(QIcon::fromTheme("qtfm",
                                   QIcon(":/images/qtfm.png")));
    setWindowTitle("QtFM");

    mBar = new QMenuBar(this);
    sBar = new QStatusBar(this);

    fileMenu = new QMenu(this);
    fileMenu->setTitle(tr("File"));

    editMenu = new QMenu(this);
    editMenu->setTitle(tr("Edit"));

    viewMenu = new QMenu(this);
    viewMenu->setTitle(tr("View"));

    navBar = new QToolBar(this);
    navBar->setWindowTitle(tr("Navigation"));

    pathEdit = new QComboBox(this);
    pathEdit->setEditable(true);
    pathEdit->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);

    mdi = new QMdiArea(this);
    mdi->setViewMode(QMdiArea::TabbedView);
    //mdi->setTabPosition(QTabWidget::South);
    mdi->setTabsClosable(true);
    mdi->setTabsMovable(true);

    backButton = new QPushButton(this);
    backButton->setText(tr("Back"));

    upButton = new QPushButton(this);
    upButton->setText(tr("Up"));

    homeButton = new QPushButton(this);
    homeButton->setText(tr("Home"));

    tileButton = new QPushButton(this);
    tileButton->setText(tr("Tile"));

    tileAction = new QAction(this);
    tileAction->setText(tr("Tile Tabs"));
    tileAction->setShortcut(QKeySequence(tr("Ctrl+S")));

    viewMenu->addAction(tileAction);

    mBar->addMenu(fileMenu);
    mBar->addMenu(editMenu);
    mBar->addMenu(viewMenu);

    //navBar->addWidget(backButton);
    //navBar->addWidget(upButton);
    //navBar->addWidget(homeButton);
    navBar->addWidget(pathEdit);
    navBar->addWidget(tileButton);

    setMenuBar(mBar);
    setStatusBar(sBar);
    addToolBar(Qt::TopToolBarArea, navBar);
    setCentralWidget(mdi);

    mimes = new MimeUtils(this);
    mimes->setDefaultsFileName(Common::readSetting("defMimeAppsFile",
                                                   MIME_APPS).toString());

    backButton->hide();
    upButton->hide();
    homeButton->hide();

    setupConnections();
    loadSettings();
    parseArgs();
}

QtFM::~QtFM()
{
    writeSettings();
}

void QtFM::newSubWindow(QString path)
{
    qDebug() << "newSubWindow" << path;
    QFileInfo info(path);
    if (!info.isDir()) { return; }

    QMdiSubWindow *subwindow = new QMdiSubWindow;
    FM *fm = new FM(mime, mimes, path);

    connect(fm, SIGNAL(newWindowTitle(QString)),
            subwindow, SLOT(setWindowTitle(QString)));
    connect(fm, SIGNAL(updatedDir(QString)),
            this, SLOT(handleUpdatedDir(QString)));
    connect(fm, SIGNAL(newPath(QString)),
            this, SLOT(handleNewPath(QString)));

    subwindow->setWidget(fm);
    subwindow->setAttribute(Qt::WA_DeleteOnClose);
    subwindow->setWindowTitle(info.completeBaseName());
    subwindow->setWindowIcon(windowIcon());
    mdi->addSubWindow(subwindow);//->setWindowState(Qt::WindowMaximized);
}

void QtFM::parseArgs()
{
    QStringList args = qApp->arguments();
    for (int i=1;i<args.count();++i) {
        if (!QFile::exists(args.at(i))) { continue; }
        newSubWindow(args.at(i));
    }
    if (mdi->subWindowList().count() == 0) {
        newSubWindow(QDir::currentPath());
    }
}

void QtFM::setupConnections()
{
    connect(tileButton, SIGNAL(released()),
            mdi, SLOT(tileSubWindows()));
    connect(tileAction, SIGNAL(triggered()),
            mdi, SLOT(tileSubWindows()));
    connect(mdi, SIGNAL(subWindowActivated(QMdiSubWindow*)),
            this, SLOT(handleTabActivated(QMdiSubWindow*)));
}

void QtFM::loadSettings()
{
    qDebug() << "load settings";
}

void QtFM::writeSettings()
{
    qDebug() << "write settings";
}

void QtFM::handleNewPath(QString path)
{
    Q_UNUSED(path)
    FM *fm = dynamic_cast<FM*>(sender());
    if (!mdi->currentSubWindow() || !fm) { return; }
    if (fm != dynamic_cast<FM*>(mdi->currentSubWindow()->widget())) { return; }
    qDebug() << "update path for tab";
    refreshPath(fm);
}

void QtFM::handleUpdatedDir(QString path)
{
    qDebug() << "handle updated dir" << path;
}

void QtFM::handleTabActivated(QMdiSubWindow *tab)
{
    if (!tab) { return; }
    FM *fm = dynamic_cast<FM*>(tab->widget());
    if (!fm) { return; }
    qDebug() << "handle tab activated" << fm->getPath();
    refreshPath(fm);
}

void QtFM::refreshPath(FM *fm)
{
    if (!fm) { return; }
    pathEdit->clear();
    pathEdit->addItems(*fm->getHistory());
    pathEdit->setCurrentIndex(0);
}
